File: Trenduity/packages/types/src/dto/card.ts
import { z } from 'zod';

/**
 * 카드 타입
 */
export const CardTypeSchema = z.enum(['ai', 'trend', 'safety', 'mobile']);
export type CardType = z.infer<typeof CardTypeSchema>;

/**
 * 카드 상태
 */
export const CardStatusSchema = z.enum(['pending', 'active', 'completed']);
export type CardStatus = z.infer<typeof CardStatusSchema>;

/**
 * 퀴즈 스키마
 */
export const QuizSchema = z.object({
  question: z.string(),
  options: z.array(z.string()),
  correctIdx: z.number(),
  explanation: z.string(),
});
export type Quiz = z.infer<typeof QuizSchema>;

/**
 * 카드 DTO
 */
export const CardDTOSchema = z.object({
  id: z.string().uuid(),
  userId: z.string().uuid(),
  date: z.string(), // ISO date
  type: CardTypeSchema,
  title: z.string(),
  tldr: z.string(),
  body: z.string(),
  impact: z.string(),
  quizzes: z.array(QuizSchema),
  status: CardStatusSchema,
  completedAt: z.string().optional(), // ISO datetime
  quizScore: z.number().min(0).max(1).optional(),
});
export type CardDTO = z.infer<typeof CardDTOSchema>;

File: Trenduity/services/bff-fastapi/app/schemas/card.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict
from datetime import datetime


class Quiz(BaseModel):
    id: str
    question: str
    options: List[str]
    correctIndex: int = Field(ge=0)
    explanation: str


class CardPayload(BaseModel):
    title: str
    tldr: str
    body: str
    impact: str
    quiz: List[Quiz] = []


class CardResponse(BaseModel):
    id: str
    user_id: str
    date: str  # ISO date
    type: str  # 'ai_tools', 'digital_safety', 'health_info'
    payload: CardPayload
    status: str  # 'pending', 'completed'
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class CardCompleteRequest(BaseModel):
    card_id: str
    quiz_answers: Optional[Dict[str, int]] = None  # { "q1": 1, "q2": 0 }


class QuizDetail(BaseModel):
    question_id: str
    is_correct: bool
    explanation: str


class QuizResult(BaseModel):
    correct: int
    total: int
    details: List[QuizDetail]


class CardCompleteResponse(BaseModel):
    points_added: int
    total_points: int
    streak_days: int
    quiz_result: Optional[QuizResult] = None
    new_badges: List[str] = []
